lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -0,0 +1,224 @@
+// maximum flexibility for optimization after removal of the nodes from the old
+// module.
+static SetVector<Operation *> computeCloneSet(ArrayRef<Operation *> roots) {
+  SetVector<Operation *> results;
+  // Get Dataflow for roots
+  dataflowSlice(roots, results);
+
+  // Get Blocks
+  SetVector<Operation *> blocks;
+  blockSlice(roots, blocks);
+  blockSlice(results.getArrayRef(), blocks);
+
+  // Make sure dataflow to block args (if conds, etc) is included
+  dataflowSlice(blocks.getArrayRef(), results);
+
+  // include the blocks and roots to clone
+  results.insert(roots.begin(), roots.end());
+  results.insert(blocks.begin(), blocks.end());
+
+  return results;
+}
+
+// Given a set of values, construct a module and bind instance of that module
+// that passes those values through.  Returns the new module and the instance
+// pointing to it.
+static hw::HWModuleOp createModuleForCut(hw::HWModuleOp op,
+                                         SetVector<Value> &inputs,
+                                         BlockAndValueMapping &cutMap,
+                                         StringRef suffix, StringRef path) {
+  OpBuilder b(op->getParentOfType<mlir::ModuleOp>()->getRegion(0));
+
+  // Construct the ports, this is just the input Values
+  SmallVector<hw::ModulePortInfo> ports;
+  for (auto port : llvm::enumerate(inputs))
+    ports.push_back(
+        {b.getStringAttr(""), hw::INPUT, port.value().getType(), port.index()});
+
+  // Create the module, setting the output path if indicated
+  std::array<NamedAttribute, 1> pathAttr = {
+      b.getNamedAttr("outputPath", b.getStringAttr(path))};
+  auto newMod = b.create<hw::HWModuleOp>(
+      op.getLoc(),
+      b.getStringAttr((getVerilogModuleNameAttr(op).getValue() + suffix).str()),
+      ports, path.empty() ? ArrayRef<NamedAttribute>() : pathAttr);
+
+  // Update the mapping from old values to cloned values
+  for (auto port : llvm::enumerate(inputs))
+    cutMap.map(port.value(), newMod.body().getArgument(port.index()));
+  cutMap.map(op.getBodyBlock(), newMod.getBodyBlock());
+
+  // Add an instance in the old module for the extracted module
+  b = OpBuilder::atBlockTerminator(op.getBodyBlock());
+  auto inst = b.create<hw::InstanceOp>(
+      op.getLoc(), ArrayRef<Type>(), ("InvisibleBind" + suffix).str(),
+      newMod.getName(), inputs.getArrayRef(), DictionaryAttr(),
+      b.getStringAttr(
+          ("__ETC_" + getVerilogModuleNameAttr(op).getValue() + suffix).str()));
+  inst->setAttr("doNotPrint", b.getBoolAttr(true));
+  b = OpBuilder::atBlockEnd(
+      &op->getParentOfType<mlir::ModuleOp>()->getRegion(0).front());
+  b.create<sv::BindOp>(op.getLoc(), b.getSymbolRefAttr(inst));
+  return newMod;
+}
+
+// Some blocks have terminators, some don't
+static void setInsertPointToEndOrTerminator(OpBuilder &builder, Block *block) {
+  if (!block->empty() && block->back().mightHaveTrait<OpTrait::IsTerminator>())
```

Comment:
The only one I'm aware of is hw.output.  Since this is HW dialect specific anyway, you might consider checking for it instead of the trait.  The trait check is slower.